### ISSUE TITLE
Reactive Streams

### ISSUE THREAD

**Issue by jroper** [netty/netty#3902]
Netty should provide a reactive streams implementation.

My reason for creating this issue is because very soon we're going to need this for Play, and I want to make sure that any efforts we make to do it ourselves are not duplicating any efforts that anyone else has started. Also, if any Netty devs want to be involved in the design/implementation, this is your chance, otherwise if you're not interested we'll just make our own library that sits on top of Netty and implements reactive streams.

**Comment by hepin1989** [comment]
@jroper make sense,vert.x implement reactive stream too.so cc @purplefox.

**Comment by normanmaurer** [comment]
@jroper while I like the idea I'm currently a bit short in time to tackle this and it is not the highest priority here. But contributions welcome of course!

**Comment by jroper** [comment]
@normanmaurer of course, I wouldn't expect you to drop everything to do this, we're fully prepared to do this ourselves.  But do you have any time to review the design or implementation?  My thoughts for the design are this (note, this is all Netty 3 terminology, I don't know how similar it is to Netty 4, but we'll probably upgrade to Netty 4 first so our impl will target Netty 4):

* Publishing - for the `Publisher`, we write a `SimpleUpstreamHandler` that publishes all message events, publishes any errors received as errors, and publishes a channel closed event as on completion.  This handles the simple case of handling the entire stream using reactive streams, and will be good for things like WebSockets.  For partial stream handling, such as handling request/response bodies, the handler will be able to be extended to intercept message events, and decide if they should be published, or if they should trigger onComplete or onError.  An extended Handler will also be able to refine/transform the type of messages published.  So a request body handler will extend this, transform the type to ChannelBuffer, and will publish onComplete when LastChunk is encountered, where it will also potentially remove/replace itself in the channel pipeline.

* Subscribing - for the `Subscriber`, we implement a `ChannelUpstreamHandler` that captures the channel when it's opened, and pushes all elements it receives to the channel, using received `WriteCompletionEvent`'s to manage backpressure.  When it receives an on completion event or error event, it will close the channel.  This will be for the simple case where the stream lifecycle is bound to the channel lifecycle.  In addition, there will be a similar handler that can be inserted into the chain, and wraps an existing channel that is passed to its constructor for publishing.  This handler won't close the channel when it's finished, rather it will make a `ChannelFuture` available for when it's finished.  This will be used, for example, for writing HTTP response bodies/request bodies, and will also offer refinement/transformation of the messages, so that things other than `MessageEvent` can be subscribed to.  For both handlers, high/low watermarks for demand will be configurable.

For the implementation, we were thinking of starting with our own library - there's no reason why the netty reactive streams implementation has to be a part of the core of netty.  If you're happy with what we do, we can transfer ownership of the repo to netty.  We'll use Java (as tempting as it is for us to use Scala :smile:), and may start with sbt, though it will be straight forward to switch to maven if it comes to transferring, and we'll also use a `com.typesafe` groupId and package name, that means we can publish it to maven central for our own use, but if/when you take over and move it to `io.netty`, you can also change the package name, and that avoids any classpath issues from having the same library with two different names.

How does that sound?

**Comment by NiteshKant** [comment]
At the risk of sounding like a sales pitch, let me introduce two libraries that are in the same space (netty + reactive streams):

- [RxNetty](https://github.com/ReactiveX/RxNetty/tree/0.5.x) This layers [RxJava](https://github.com/ReactiveX/RxJava) over netty to provide client and server abstractions for various protocols. Reactive streams effort was started by developers from RxJava and hence they are very similar abstractions. [RxJava 2.0](https://github.com/ReactiveX/RxJava/issues/2450) is intended to be aligned more closely to reactive stream or Java 9 flow.

- [Reactive IPC](https://github.com/reactive-ipc/reactive-ipc-jvm): Based on the learnings from RxNetty, and developments around reactive streams, we decided to start a new library which is very similar to RxNetty but the core is based on reactive streams as opposed to RxJava. This enables the ability to layer technologies like RxJava on top of the core client/server abstractions based on reactive streams. The goals and motivations can be found in [this issue](https://github.com/reactive-ipc/reactive-ipc-jvm/issues/1). This library is at a very nascent stage and we are just starting to discuss the core abstractions.

It sounds like you are having similar aspirations/requirements as these libraries, so I thought it will make sense to introduce them here and see if we can cross-pollinate ideas, if not collaborate on the same library.

**Comment by hepin1989** [comment]
@jroper I think you should start with netty4,in fact the netty4's thread model is more flexible ,and you could control the read/accept ,which means back pressure.and you could trigger a next read/accept when there is a request from downstream.

I just implement akka-io-netty using netty4.1 and it nearly implement all the command of akka-io,so I think  it would be better to using netty4 as a start.

btw you could new a channel and then register/deregister it to event loop too.for the channel ,I think that should we make it a BidiFlow?

for the separate concern,you may want to implement an akka-io-netty extension and then implement the stream based on it ,just like what akka does with the the current implement of the akka io.
https://github.com/akka/akka/tree/release-2.3-dev/akka-stream/src/main/scala/akka/stream/impl/io

I think you should not use the bootstrap/serverbootstrap ,then you get more control of it.

**Comment by hepin1989** [comment]
@jroper I push our code here
https://github.com/qgame/engine/tree/master/framework/akka-io-netty/
you may want to take a look.

**Comment by jroper** [comment]
Actually the reason we want this is that Play is moving to Akka streams for the IO library that we're exposing in our API. So we need something to bridge Akka streams on to Netty, and a direct reactive streams implementation is the most logical choice, since Akka streams of course already supports reactive streams. My concern about layering something like RxJava in between is that performance may suffer.

**Comment by hepin1989** [comment]
@jroper So still have netty based server there?I think only akka-http based one will contains in the next Play release,and you will left the netty one to the community.

Seems like you have to implement the directly one.

**Comment by jroper** [comment]
@hepin1989 The roadmap has been updated, Play 2.5 is now going to use Akka streams, and will still use Netty.

**Comment by normanmaurer** [comment]
@jroper sounds good. I would for sure be happy to review.

**Comment by jroper** [comment]
Ok, so I'm ready for a firest review of the publisher approach.  You can also review the publisher if you like, though the more important thing at this stage is the approach.  You can read the relevant javadocs here:

https://github.com/typesafehub/netty-reactive-streams/blob/master/src/main/java/com/typesafe/netty/HandlerPublisher.java

So, `HandlerPublisher` is similar to `SimpleInboundChannelHandler`, it selects which messages to publish based on the passed in type. It's up to an earlier handler to control how that should work.  The end of stream is signalled either via `channelInactive`, or via reading a `HandlerPublisher.COMPLETED` message.  If the subscriber cancels, this is signalled by publishing a `disconnect`, which an earlier handler of course can intercept if it wants to continue handling the stream.  Any exceptions received cause the subscriber to be immediately given an error, any buffered messages are released.

There are a few things I'm not sure about.  What happens to the executor when the handler is removed from the pipeline?  If, concurrently, the subscriber is signalling demand, ie, submitting an event to the `ChannelHandlerContext` executor, is it likely to get an exception?  Is it valid to use the executor for the `ChannelHandlerContext` after the handler is removed from the pipeline?  How about after the channel has been closed?  The problem is, the handler could be removed from the pipeline, eg because it's only handling a partial stream, and the channel could even be closed, but the subscriber may still have buffered elements to consume.

From looking at the netty code, it seems that it's always ok to use the channel handler contexts executor, if you didn't provide your own, it just uses the channels, and this is never cleared, no state ever changes, as far as I can see.  Given that, it would probably also be safe to replace the `COMPLETE` message with simply removing the handler from the pipeline, and handle it in the `handlerRemoved` callback.  Would that be a better way to do things?

**Comment by jroper** [comment]
I just tried out using removing the handler as a mechanism, the reason it doesn't work is that removing the handler is always done synchronously, while publishing a read event may be done asynchronously, that means you can't publish a read event and then remove the handler, because the handler may be removed before it receives the read event, and so the read event will be lost.

I'm wondering if using `channelInactive` would be a good idea instead of `COMPLETE`?  This would simplify things even more.  I just don't know if it's considered good practice in Netty to signal events that haven't actually happened.

**Comment by jroper** [comment]
Ok, I've now got this to a point where I'm happy that it's working and feature complete.  Both the publisher and subscriber TCKs pass.  I have made no attempt to optimise performance, however, the performance won't be bad, it is completely non blocking.  When a context is available (once `handlerAdded` is invoked), all events that didn't come from Netty are sent to the handlers context.  Prior to that, AtomicReferences are used to manage state.  Feel free to review at your leisure:

https://github.com/typesafehub/netty-reactive-streams

I've also created a second project which is not yet complete, but it provides reactive streams support for http, by creating new sub interfaces of `HttpRequest` and `HttpResponse` called `StreamedHttpRequest` and `StreamedHttpResponse` which provide `Publisher<HttpContent>` which can (or if AUTO_READ is false, must) be subscribed to to handle the request/response bodies, and a `HttpStreamClientHandler` and `HttpStreamServerHandler` that handles these new types, as well as `FullHttpRequest` and `FullHttpResponse`.  In addition to using this handler, you can also then insert additional publisher/subscriber handlers into the pipeline to handle `HttpRequest`/`HttpResponse`, and then you can create a `Processor<HttpRequest, HttpResponse>` that is an HTTP client, and an HTTP server that uses a `Processor<HttpRequest, HttpResponse>` to handle requests.  I've actually done this in the tests, and run the reactive streams TCK for `Processor` against a client and server setup, and it works.  As I said, this part is not yet complete (doesn't handle expect 100 continue yet for example), but feel free to review if you're interested.

**Comment by jroper** [comment]
A further update on this, I've ported Play to use this (in a branch at the moment because some other things need to be done first), and our tests are passing.  It won't be too long before we get this into Play's master branch, and then it'll be run by our performance test infrastructure, which should be interesting.

**Comment by normanmaurer** [comment]
@jroper awesome! Let us know once you have the results

**Comment by jroper** [comment]
A further update on this, this support is now in the Play master branch, handling request bodies, response bodies, and WebSockets using reactive streams.  Additionally, we've done some initial work in putting support for this into async-http-client, and the good news is that it was very easy to integrate - this I think validates the API design, since now we have two separate libraries using it successfully.

Performance wise, so far we've seen a very small downtick in Play's performance - this is not wholly unexpected, the approach we've taken is to use reactive streams for *everything*, the incoming connections are exposed as a `Publisher<Channel>`, the requests are then handled by a `Processor<HttpRequest, HttpResponse>`, and then when handling request/response bodies, we use `Publisher<HttpContent>`.  We then use Akka streams to produce/consume these streams, and what this is doing is introducing additional asynchronous boundaries, where control goes from a Netty thread to an Akka thread, that didn't previously exist.  In Play we have always dispatched all requests to an Akka thread, but only once, the nature of reactive streams, and Akka's implementation of it, introduces a minimum of an additional two thread switches that didn't previous exist.  The cost of these significantly outweighs any performance gains made by upgrading to Netty 4.  So, as a first step, we'll switch to implementing ordinary channel handlers for handling requests, and just use reactive streams for the request/response bodies, and this should get our performance back to what it was (or maybe even improve it).

**Comment by normanmaurer** [comment]
Thanks James!

**Comment by hepin1989** [comment]
@jroper for the accept connection,I think foreachParallel should help.

**Comment by jroper** [comment]
Hi all,

So, netty-reactive-streams is now in Play 2.5.0, and is stable, performs well, being used in production, and we're quite happy with it.

The question is where to next.  I don't have any problem continuing to maintain it, that said, I think it'd be a great feature to have as a module in the main netty repo, people wanting to use reactive streams with Netty will probably be much happier consuming it if it was part of netty itself than something coming from Lightbend.

What I've written has two parts - first is the reactive streams module which has a publisher handler and a subscriber handler that are able to turn a channel into a reactive streams publisher/subscriber - this is the main reactive streams support which I think should definitely be in netty.

The second is a module with http specific support for reactive streams - it's somewhat similar in purpose to the HTTP object aggregators, it allows developers to work at a higher level by ensuring that only HttpMessages go down the pipeline so users don't have to handle HttpContent messages in the pipeline themselves, but in addition to producing FullHttpRequest/FullHttpResponse (which it does when there's no body), it will also produce StreamedHttpRequest/StreamedHttpResponse, which allows you to consume the request/response body using reactive streams.  It also has support for pipelining and in the server side, websockets.  Though very useful, this module is not as important to be in netty itself.  We're using it in Play and will continue to maintain and improve it regardless.  So I don't have any strong opinion about where it should live.

**Comment by trustin** [comment]
I didn't really have a chance to take a close look yet. However, a quick look makes me like the simplicity of reactive stream and your implementation.

@normanmaurer, @nmittler, @Scottmitch - could you review his work and let me know what you think? (pulling reactive stream in, streaming HTTP content using reactive stream, etc)

**Comment by jlouvel** [comment]
@jroper We have started building the prototype of [Restlet Framework NG](https://github.com/restlet/restlet-framework) on top of Netty 4.1 and Reactive Streams using your library. So far it works nicely and looks like a perfect match for our needs!

@trustin @normanmaurer I would love to have it part of the Netty project to facilitate collaboration and contributions. Happy to contribute to make it happen.

**Comment by jroper** [comment]
@jlouvel I assume you're using both the `netty-reactive-streams` module as well as `netty-reactive-streams-http`.  What's your opinion on whether `netty-reactive-streams-http` belongs in Netty or not?

**Comment by jlouvel** [comment]
@jroper Yes I'm using both modules and think it would make sense to have both of them in Netty at this stage. But I'm not very advanced in term of implementation as I haven't added support for HTTP body streaming yet (series of HttpContent events).

**Comment by normanmaurer** [comment]
Let's close this. We could have users contrib it to netty-contribs tho.
